[AutoFill Debugging] Add an interaction type to scroll by a given distance

whsieh opened the pull request:
#### 78f799e6ecc67e9bc5a6dc35028edca1653a7c4f
<pre>
[AutoFill Debugging] Add an interaction type to scroll by a given distance
<a href="https://bugs.webkit.org/show_bug.cgi?id=302728">https://bugs.webkit.org/show_bug.cgi?id=302728</a>
<a href="https://rdar.apple.com/164990607">rdar://164990607</a>

Reviewed by Abrar Rahman Protyasha and Megan Gardner.

Add support for a new interaction type, `_WKTextExtractionActionScrollBy`. This interaction takes:

-   A scrolling offset delta, in the form of a new property on `_WKTextExtractionInteraction` (this
    must be non-zero)

-   A node identifier representing a subscrollable container (or a node underneath a subscrollable
    container); if unspecified, this scrolls the page.

* LayoutTests/fast/text-extraction/debug-text-extraction-interactions-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-interactions.html:

Augment this existing test to cover the change.

* Source/WebCore/page/EventHandler.h:

Move this helper into the public section, so that we can call it below.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::scrollBy):
(WebCore::TextExtraction::handleInteraction):
(WebCore::TextExtraction::interactionDescription):
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _convertToWebCoreInteraction:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionInteractionOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::performTextExtractionInteraction):

Canonical link: <a href="https://commits.webkit.org/303235@main">https://commits.webkit.org/303235@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ede421c8fafb2f2bd96a9d956bb900b40bcace1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4192 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42700 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139209 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/83564 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7734710f-85e4-446a-a062-fee4bfccfb4e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133570 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4118 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3954 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100676 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/68109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/24478308-967c-4247-9569-4910b047c98f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134646 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117968 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81446 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b5d3b18a-7f0e-4586-bb0e-66df71ee3f94) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2864 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/702 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82431 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111592 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141855 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3861 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36671 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109039 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3942 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3415 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109196 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27671 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2953 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114248 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57071 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3914 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32665 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3746 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/67362 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4006 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3875 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -1330,6 +1330,24 @@ static void highlightText(Page& page, std::optional<NodeIdentifier>&& identifier
     return completion(true, { });
 }
 
+static void scrollBy(Page& page, std::optional<NodeIdentifier>&& identifier, FloatSize scrollDelta, CompletionHandler<void(bool, String&&)>&& completion)
```
Meta comment: it seems like the completion arguments `(bool, String)` are very `Expected<std::monostate, String>` like.

whsieh replied:
Yeah... I'd initially considered it when writing this code, but there are some cases where we return `true` (success) but _additionally_ surface text describing how the action was handled.

aprotyas commented on Source/WebCore/page/text-extraction/TextExtraction.cpp:
```diff
@@ -1476,6 +1494,12 @@ void handleInteraction(Interaction&& interaction, Page& page, CompletionHandler<
+            return completion(false, "Scroll delta is zero"_s);
+
+        return scrollBy(page, WTFMove(interaction.nodeIdentifier), interaction.scrollDelta, WTFMove(completion));
+    }
```
nit: I think you don't need a new scope here, since you aren't defining any variable

whsieh replied:
Good point, fixed!

aprotyas commented on Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
```diff
@@ -441,6 +443,11 @@ - (void)paste:(id)sender;
+    if (auto scrollDelta = options->scrollDelta) {
+        auto [x, y] = *scrollDelta;
+        [interaction setScrollDelta:CGSizeMake(x, y)];
+    }
```
optional, saves a line 😄 
```suggestion
    if (auto scrollDelta = options->scrollDelta)
        [interaction setScrollDelta:std::apply(CGSizeMake, *scrollDelta)];
```

whsieh replied:
Oh neat, didn't realize that was a thing!

whsieh committed:
[AutoFill Debugging] Add an interaction type to scroll by a given distance
https://bugs.webkit.org/show_bug.cgi?id=302728
rdar://164990607

Reviewed by Abrar Rahman Protyasha and Megan Gardner.

Add support for a new interaction type, `_WKTextExtractionActionScrollBy`. This interaction takes:

-   A scrolling offset delta, in the form of a new property on `_WKTextExtractionInteraction` (this
    must be non-zero)

-   A node identifier representing a subscrollable container (or a node underneath a subscrollable
    container); if unspecified, this scrolls the page.

* LayoutTests/fast/text-extraction/debug-text-extraction-interactions-expected.txt:
* LayoutTests/fast/text-extraction/debug-text-extraction-interactions.html:

Augment this existing test to cover the change.

* Source/WebCore/page/EventHandler.h:

Move this helper into the public section, so that we can call it below.

* Source/WebCore/page/text-extraction/TextExtraction.cpp:
(WebCore::TextExtraction::scrollBy):
(WebCore::TextExtraction::handleInteraction):
(WebCore::TextExtraction::interactionDescription):
* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _convertToWebCoreInteraction:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.mm:
* Tools/TestRunnerShared/UIScriptContext/Bindings/UIScriptController.idl:
* Tools/TestRunnerShared/UIScriptContext/UIScriptController.h:
* Tools/TestRunnerShared/UIScriptContext/UIScriptControllerShared.cpp:
(WTR::toTextExtractionInteractionOptions):
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::performTextExtractionInteraction):

Canonical link: https://commits.webkit.org/303235@main
